Split from gnome-initial-setup

wjt opened the pull request:
This PR removes all unused gnome-initial-setup pages, and also drops the language page, which was previously used on standalone `eosinstaller` images. Other gnome-initial-setup source files which are *entirely* unused are removed, too. There is still some dead code in the remaining source files – in particular, support for switching locale on the fly, which is no longer needed – but the branch was getting big enough as it is!

I manually merged some fixes to the remaining files from the intervening 2 years of gnome-initial-setup development. I also did some cleanup I found along the way.

https://phabricator.endlessm.com/T17082

wjt committed:
Remove language page from page list

We will arrange to always be launched from the real gnome-initial-setup
after the language page, with the locale adjusted as appropriate.

https://phabricator.endlessm.com/T17082

wjt committed:
Remove unused UIDIR definition

This is used in g-i-s, but not here.

wjt committed:
Remove unused gnome-initial-setup files

These are files which are entirely unused by anything in
gnome-image-installer/ or eos-installer-data/. There is certainly more
unused code in gnome-initial-setup/ but we will address that piece by
piece.

https://phabricator.endlessm.com/T17082

wjt committed:
configure: Remove gnome-initial-setup dependencies

https://phabricator.endlessm.com/T17082

wjt committed:
Drop unused accountsservice code and dependency

https://phabricator.endlessm.com/T17082

wjt committed:
Remove unneeded keyring manipulation code

eos-installer is now launched either by the "real" FBE or from within a
user session in all cases.

https://phabricator.endlessm.com/T17082

wjt committed:
driver: remove unused hide_window function

wjt committed:
page: remove stray get_action_widget prototype

This function doesn't exist.

wjt committed:
driver: don't set user locale

The locale for the current user, be it gnome-initial-setup or live, will
have already been configured by the FBE.

wjt committed:
README: update for g-i-s-ectomy

https://phabricator.endlessm.com/T17082

wjt committed:
Remove FBE session files

eos-installer is now launched from the real FBE, even on standalone
eosinstaller images.

https://phabricator.endlessm.com/T17082

wjt committed:
Install .desktop file to /usr/share/applications

This was the only remaining file installed to
/usr/share/gdm/greeter/applications. It makes slightly more conceptual
sense to install it to the standard directory, since it can be launched
from (live) user sessions as well as the FBE, though of course this
requires changes to the corresponding code in gnome-initial-setup.

(“Would you jump off a cliff if your friends did?” notwithstanding, this
new arrangement matches the corresponding setup in Fedora.)

https://phabricator.endlessm.com/T17082

wjt committed:
finished: remove spurious message about reading eos-image-version

The eos-image-version xattr is stored on the real root directory of the
root filesystem. On converted systems, this is `/sysroot`, while on
unconverted systems, this is `/`. We check them in that order.

Previously, when run on a converted system, we would always print a
message about being unable to read this attribute from `/sysroot`. But
this is expected! It's only a problem if we can't read it from `/`
either; so, only print a message in that case.

https://phabricator.endlessm.com/T17082

wjt committed:
Split GisDriverMode to its own file

https://phabricator.endlessm.com/T14616

wjt committed:
Pass GisDriverMode into GisAssistant

This is functionally identical since, previously, enable_controls <=>
mode == GIS_DRIVER_MODE_EXISTING_USER. We will introduce more UI &
behaviour changes based on the mode, and 'enable_controls' isn't
descriptive enough.

https://phabricator.endlessm.com/T14616

wjt committed:
Allow going Back from first page when run from FBE

This allows the user to get back to the Try/Reformat choice page (in the
combined USB/ISO case) or the Language page (in the eosinstaller image
case).

It's enough to just make the process exit cleanly, since in FBE mode
there is no other way this can happen: on the final page, there's just a
Power Off button, and this will (indirectly) cause the process to be
killed. The FBE has been modified to take the correct action when this
process exits cleanly.

https://phabricator.endlessm.com/T14616

wjt committed:
driver: don't inhibit idle

All images now use the real gnome-initial-setup in the FBE, which takes
care of this for us. We also care a bit less about inhibiting idle
within a user session. So, just don't do this at all.

https://phabricator.endlessm.com/T17082

wjt committed:
assistant: Remove progress indicator

This is consistent with gnome-initial-setup, where commit
ecc7ff533a39374a60f6cf9c10fdfb9b81c2d006 removed this back in GNOME 3.24
/ Endless OS 3.2.0.

wjt committed:
assistant: Remove unused enum / "direction" parameter

(gnome-initial-setup 094b675ec009fe93a790c7f565567edcf0610f04)

wjt committed:
assistant: Remove next-page signal

(gnome-initial-setup c97ae3ea8911724f19f3017d817d855ba282c02e)

wjt committed:
disktarget: Convert to composite widget template

wjt committed:
diskimage: Convert to composite widget template

wjt committed:
install: convert to composite widget template

wjt committed:
finished: convert to composite widget template

wjt committed:
gis-page: Remove get_builder vfunc

All pages now use composite widget templates

(gnome-initial-setup 0e06bbcbc6a5ed39dba5040c8e768de56a0e46b7)

wjt committed:
main: remove unneeded gis_page_get_type() call

wjt committed:
window: fix icon-name

smspillaz commented on gnome-initial-setup/gis-page.h:
```diff
@@ -78,7 +78,6 @@ const char *       gis_page_get_forward_text (GisPage *page);
 void         gis_page_set_forward_text (GisPage *page, const char *text);
 gboolean     gis_page_get_complete (GisPage *page);
 void         gis_page_set_complete (GisPage *page, gboolean complete);
-GtkWidget *  gis_page_get_action_widget (GisPage *page);
```
Does this need to be sent upstream too?

wjt replied:
Good point. https://bugzilla.gnome.org/show_bug.cgi?id=793833

smspillaz commented on eos-installer-data/com.endlessm.Installer.desktop.in.in:
```diff
@@ -7,3 +7,4 @@ Type=Application
 StartupNotify=true
 Categories=GNOME;GTK;System;
 OnlyShowIn=GNOME;
+NoDisplay=true
```
I don't quite understand why we have a `NoDisplay` attribute here - wouldn't we want the launcher for the installer to be visible in things like the search? Or I guess this attribute is dynamically removed on live sessions and not for installed ones, right?

Also, what happens if a user accidentally launches the installer on an installed session? Should the installer detect that it is not in a live boot and bail out accordingly? (I suppose we already had this problem when the desktop file was in /usr/share/gdm/greeter/applications, though it came to mind just now).

wjt replied:
> I don't quite understand why we have a NoDisplay attribute here - wouldn't we want the launcher for the installer to be visible in things like the search? Or I guess this attribute is dynamically removed on live sessions and not for installed ones, right?

Right: we only want the reformatter to be visible in live sessions. This is the purpose of `NoDisplay=true` here. There is a [corresponding patch in the g-i-s PR](https://github.com/endlessm/gnome-initial-setup/pull/227/commits/1d67e2874a2c8e56880f6ddb0903f19ec68a5e71) which removes `NoDisplay=true` on live boots. Previously, it was hidden from non-live sessions by putting into a non-standard path.

wjt replied:
> Also, what happens if a user accidentally launches the installer on an installed session? Should the installer detect that it is not in a live boot and bail out accordingly? (I suppose we already had this problem when the desktop file was in /usr/share/gdm/greeter/applications, though it came to mind just now).

![screenshot from 2018-02-26 09-38-15](https://user-images.githubusercontent.com/86760/36663318-1b80618e-1ad9-11e8-822b-a054562a711d.png)

smspillaz replied:
Heh, fair enough. I guess it'd be nice have a nicer error for what the problem actually is, but that's outside the scope of this PR.

smspillaz commented on gnome-image-installer/pages/finished/gis-finished-page.c:
```diff
@@ -134,42 +134,54 @@ toggle_leds (GisPage *page)
+  g_set_error (error, G_IO_ERROR, g_io_error_from_errno (errsv),
+               "Error examining " EOS_IMAGE_VERSION_XATTR " on %s: %s",
+               path, g_strerror (errsv));
+  return NULL;
```
Since we're using `g_autoptr` above, it'd be nice to avoid the goto error stuff here - to avoid the `g_set_error` duplication you can wrap it in a function like `error_about_invalid_xattr (GError **);`, which just sets the error and returns.

wjt replied:
I don't think adding an extra function improves clarity, but I have eliminated the `goto` in a fixup commit. See what you think. The same code and fix is [present in g-i-s](https://github.com/endlessm/gnome-initial-setup/pull/227/commits/4a9a44d1697109ebde3a08527b2db30ae6832114) so I'll apply the same change there if this is preferable.

smspillaz replied:
Fair enough. Its more of a personal preference and not important.

wjt replied:
I squashed the fixup commit out of habit, sorry :(

smspillaz commented on gnome-image-installer/pages/finished/gis-finished-page.c:
```diff
@@ -134,42 +134,54 @@ toggle_leds (GisPage *page)
 {
-  g_autofree gchar *image_version = get_image_version (EOS_IMAGE_VERSION_PATH);
+  g_autoptr(GError) error1 = NULL;
+  g_autoptr(GError) error2 = NULL;
```
Maybe `error_image_version_path` and `error_image_version_alt_path`?

wjt replied:
How about `error_sysroot` and `error_root`?

smspillaz replied:
That works!

smspillaz commented on gnome-image-installer/pages/finished/gis-finished-page.c:
```diff
@@ -134,42 +134,54 @@ toggle_leds (GisPage *page)
+    {
+      g_warning ("%s", error1->message);
+      g_warning ("%s", error2->message);
+    }
```
What errors can be returned here? Some more context would be nice, eg `g_warning ("Error reading standard image version path: %s")`

Also, should we show a warning about being able to read from the standard image version path if we were able to successfully read the alternative one?

wjt replied:
> What errors can be returned here? Some more context would be nice, eg g_warning ("Error reading standard image version path: %s")

There is only one error, which you can see by scrolling up to the function directly above this one:

```c
  g_set_error (error, G_IO_ERROR, g_io_error_from_errno (errsv),
               "Error examining " EOS_IMAGE_VERSION_XATTR " on %s: %s",
               path, g_strerror (errsv));
```

It includes all of the context you describe.

> Also, should we show a warning about being able to read from the standard image version path if we were able to successfully read the alternative one?

The whole point of this change is not to do that! Here is the commit message:

> page-util: remove spurious message about reading eos-image-version
> 
> The eos-image-version xattr is stored on the real root directory of the
> root filesystem. On converted systems, this is `/sysroot`, while on
> unconverted systems, this is `/`. We check them in that order.
> 
> Previously, when run on a converted system, we would always print a
> message about being unable to read this attribute from `/sysroot`. But
> this is expected! It's only a problem if we can't read it from `/`
> either; so, only print a message in that case.

smspillaz replied:
Makes sense (looking at the error above)

smspillaz commented on gnome-initial-setup/gis-driver.h:
```diff
@@ -52,11 +53,6 @@ struct _GisDriverClass
-  GIS_DRIVER_MODE_NEW_USER,
-  GIS_DRIVER_MODE_EXISTING_USER,
-} GisDriverMode;
-
```
This patch can probably be upstreamed too

wjt replied:
It's not necessary. I split this to a different file so it could be referenced from `gis-assistant.h` without introducing a circular dependency between `gis-assistant.h` and `gis-driver.h`.

smspillaz replied:
Makes sense

smspillaz commented on gnome-initial-setup/gis-driver.h:
```diff
@@ -71,8 +71,7 @@ void gis_driver_add_page (GisDriver *driver,
 
-GisDriver *gis_driver_new (GisDriverMode mode,
-                           gboolean      inhibit_idle);
+GisDriver *gis_driver_new (GisDriverMode mode);
```
Do any corresponding polkit rules files need to be removed here as well? (I only ask because I had to override some polkit rules in order to make inhibit idle work on a different project).

wjt replied:
No, this inhibits idle by talking to gnome-session. System services talk directly to logind to inhibit idle, and I see there is a polkit action for that.

smspillaz replied:
Makes sense

smspillaz commented on gnome-initial-setup/gis-assistant.c:
```diff
@@ -103,17 +102,38 @@ gis_assistant_switch_to (GisAssistant          *assistant,
+  while (!should_show_page (l)) {
+    l = l->next;
+  }
+  return GIS_PAGE (l->data);
```
What happens if `l == NULL` here due to running out of pages to show?

wjt replied:
Indeed! The simple answer is that this doesn't happen, because the forward button is only shown if we are not on the last page (and `should_show_page` returns TRUE for the last page). The same bomb appears in g-i-s upstream so I suppose I'll go fix it there.

smspillaz replied:
Sounds good

wjt replied:
https://bugzilla.gnome.org/attachment.cgi?id=368925&action=edit

smspillaz commented:
Reviewed, other than those questions. It'd be good to get someone more familiar with GtkBuilder to review the conversion to composite widget templates, that's not an area I'm particularly familiar with.

smspillaz commented:
I'm happy with the responses where indicated

wjt committed:
assistant: don't crash if we run out of pages

In the previous implementation of these loops, if (for some reason) all
pages in the relevant direction were hidden, the `l = l->next` or `l =
l->priv` lines would dereference a NULL pointer. In practice this does
not happen, and it would represent a bug in the application if it did;
nonetheless, it's neater to log a critical than to abort.

https://bugzilla.gnome.org/show_bug.cgi?id=793833

ramcq commented:
I am no GtkBuilder guru but I am also kind of relaxed about these changes - the code changes in those few patches seem sane, so if it builds, runs and looks OK, I'm not going to review every last XML diff.
